### DISCUSSION TITLE
Open create modal with fill data from widget

### DISCUSSION THREAD

**Discussion by ashleyhood** [filamentphp/filament#2917] [General]
First I want to say how great Filament is. Thanks for all who have made this great package and the community that means I can find most of the answers to my questions without asking :smile: 

I am using filament-calendar plugin with a simple (modal) resource. I would like to be able to click on an event on the calendar and have it open the modal with the form pre-filled with the resource model. Is this possible? 

I have extended the filament-calendar and added the dataClick event. I would also like to be able to open the create modal when a date in the calendar is clicked.

**Comment by danharrin** [comment]
Tbh it's probably best if you ask this question in the plugin repository instead of there, then you will get an answer from the creator.

But I can help you a bit, but sending example code for how you use the Filament modal component:

```html
<form id="my-calendar-modal" wire:submit.prevent="livewireSubmitMethodHere">
    <x-filament::modal>
        <x-slot name="header">
            Modal heading
        </x-slot>

        Form components here

        <x-slot name="actions">
            <button type="submit">
                 Submit form
            </button>
        </x-slot>
    </x-filament::modal>
</form>
```

Store the active calendar item in a public Livewire property, then open this modal by dispatching an `open-modal` event with `{ id: 'my-calendar-modal' }`. You can use the form builder to render the form, call `->fill()` on the form to fill it with data from the calendar item.

**Reply by ashleyhood** [reply]
Thanks for the quick reply @danharrin you are a super hero!

The problem could be extended to any widget that is added to the header/footer of a resource page. Would it be possible to open the edit modal (just as if you had clicked edit on the resource table row) from a widget on the resource page.

If i don't have to create a whole new modal and form but instead just activate the edit modal and form that is used on the resource table row. This is the default edit modal for a [simple resources](https://filamentphp.com/docs/2.x/admin/resources/getting-started#simple-modal-resources).

**Reply by danharrin** [reply]
Ah ok, I didn't realise you wanted to open an existing resource modal.

In that case, you could emit a Livewire event from the widget to open table action. On the List page, register the listener and call `$this->mountTableAction('create')` or edit.

**Reply by ashleyhood** [reply]
Brilliant @danharrin :tada: 

That works exactly as I had imagined. Filament just keeps getting better and better to use as I understand it more and more.

Thanks for your help.

**Comment by ashleyhood** [comment]
Is there a way to populate the form data in the create modal?

Here is what I have so far:
```php
class ManageAppointments extends ManageRecords
{
    protected static string $resource = AppointmentResource::class;

    protected function getActions(): array
    {
        return [
            Actions\CreateAction::make(),
        ];
    }

    protected function getHeaderWidgets(): array
    {
        return [
            CalendarWidget::class,
        ];
    }

    protected function getListeners()
    {
        return ['onEventClicked', 'onDateClick'];
    }

    public function onEventClicked($data)
    {
        $this->mountTableAction('edit', $data['id']);
    }

    public function onDateClick($data)
    {
        $startDate = Carbon::parse($data['date']);
        $this->cachedActions['create']->afterFormFilled(fn (Component $livewire) => $livewire->mountedActionData['started_at'] = $startDate);

        $this->mountAction('create');
    }
}
```
Is this the best way to add data to the create action form?

I am able to add the `$startDate` but when i use `DateTimePicker::make('started_at')` in `AppointmentResource::class` it is changing to the local browser time and not respecting the set timezone.

**Reply by ashleyhood** [reply]
I have found a solution that works and feels like it is the right approach:

```php
class ManageAppointments extends ManageRecords
{
    protected static string $resource = AppointmentResource::class;

    protected function getActions(): array
    {
        return [
            Actions\CreateAction::make(),
        ];
    }

    protected function getHeaderWidgets(): array
    {
        return [
            CalendarWidget::class,
        ];
    }

    protected function getListeners()
    {
        return ['onEventClicked', 'onDateClick'];
    }

    public function onEventClicked($data)
    {
        $this->mountTableAction('edit', $data['id']);
    }

    public function onDateClick($data)
    {
        $startDate = Carbon::parse($data['date']);

        $this->getCachedAction('create')->mountUsing(fn (ComponentContainer $form) => $form->fill(['started_at' => $startDate]));

        $this->mountAction('create');
    }
}
```

**Reply by danharrin** [reply]
Perfect!
